pkg/controller/deployment/util.go

```diff
@@ -0,0 +1,119 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package deployment
+
+import (
+	"fmt"
+	"strconv"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/apis/extensions"
+	deploymentutil "k8s.io/kubernetes/pkg/util/deployment"
+	"k8s.io/kubernetes/pkg/util/integer"
+)
+
+func getDesiredReplicasAnnotation(rs *extensions.ReplicaSet) (int32, bool) {
+	desiredString, ok := rs.Annotations[deploymentutil.DesiredReplicasAnnotation]
+	desired, err := strconv.Atoi(desiredString)
+	if !ok || err != nil {
+		if err != nil {
+			glog.Warning("Cannot convert the value %q with annotation key %q for the replica set %q",
+				desiredString, deploymentutil.DesiredReplicasAnnotation, rs.Name)
+		}
+		return int32(0), false
+	}
+	return int32(desired), ok
+}
+
+func getMaxReplicasAnnotation(rs extensions.ReplicaSet) (int32, bool) {
+	maxReplicasString, ok := rs.Annotations[deploymentutil.MaxReplicasAnnotation]
+	maxReplicas, err := strconv.Atoi(maxReplicasString)
+	if !ok || err != nil {
+		if err != nil {
+			glog.Warning("Cannot convert the value %q with annotation key %q for the replica set %q",
+				maxReplicasString, deploymentutil.MaxReplicasAnnotation, rs.Name)
+		}
+		return int32(0), false
+	}
+	return int32(maxReplicas), true
+}
+
+func setReplicaAnnotations(rs *extensions.ReplicaSet, desiredReplicas, maxReplicas int32) {
+	if rs.Annotations == nil {
+		rs.Annotations = make(map[string]string)
+	}
+	rs.Annotations[deploymentutil.DesiredReplicasAnnotation] = fmt.Sprintf("%d", desiredReplicas)
+	rs.Annotations[deploymentutil.MaxReplicasAnnotation] = fmt.Sprintf("%d", maxReplicas)
+}
+
+// maxUnavailable returns the maximum unavailable pods a rolling deployment can take.
+func maxUnavailable(deployment extensions.Deployment) int32 {
+	if !deploymentutil.IsRollingUpdate(&deployment) {
+		return int32(0)
+	}
+	// Error caught by validation
+	_, maxUnavailable, _ := deploymentutil.ResolveFenceposts(&deployment.Spec.Strategy.RollingUpdate.MaxSurge, &deployment.Spec.Strategy.RollingUpdate.MaxUnavailable, deployment.Spec.Replicas)
+	return maxUnavailable
+}
+
+// maxSurge returns the maximum surge pods a rolling deployment can take.
+func maxSurge(deployment extensions.Deployment) int32 {
+	if !deploymentutil.IsRollingUpdate(&deployment) {
+		return int32(0)
+	}
+	// Error caught by validation
+	maxSurge, _, _ := deploymentutil.ResolveFenceposts(&deployment.Spec.Strategy.RollingUpdate.MaxSurge, &deployment.Spec.Strategy.RollingUpdate.MaxUnavailable, deployment.Spec.Replicas)
+	return maxSurge
+}
+
+// estimateProportion will estimate the proportion of the provided replica set using the recorded
+// total-replicas annotation on it, the replica count that can be added, the total replicas the
+// deployment can have currently, and the total replicas added so far.
+func estimateProportion(rs *extensions.ReplicaSet, replicasToAdd, maxReplicas, totalAdded int32) int32 {
```

Comment:
Some nits on naming:  Since the purpose of this function is to return the number of replicas to add to the RS lets name it something that reflects that.  Maybe getReplicasToAdd().

Also, both the RS and Deployment have Desire replicas and Replicas to Add so names like `replicasToAdd ` are confusing.  When we are in contexts where it could refer to either, we should name the variable to be unambiguous.  The 

getRSReplicasToAdd(rs *extensions.ReplicaSet, deploymentTotalReplicasToAdd, deploymentTotalObservedReplicas, deploymentReplicasAlreadyAdded int32)

If you think the names are too long, feel free to truncate them and well document what they should be.